### PR TITLE
Bugfix/3169 stack fix

### PR DESCRIPTION
**Pull request by davidnich**

### CODE REVIEW

**Comment by davidnich** [comment]
allows the jni module to make exceptions on demand (no cost with no exception) like:
```
System exception: ../test/jni.qtest:127: NO-FUNCTION: function name 'does_not_exist' does not exist
call stack:
  org.qore.jni.QoreJavaApi.callFunction0() called at QoreJavaApi.java:-2 (Java builtin code)
  org.qore.jni.QoreJavaApi.callFunction() called at QoreJavaApi.java:15 (Java user code)
  org.qore.jni.test.QoreJavaApiTest.callFunctionTest() called at QoreJavaApiTest.java:94 (Java user code)
  QoreJavaApiTest::callFunctionTest() called at ../test/jni.qtest:127 (Qore builtin code)
  Main::exceptionStackTest() called at /Users/david/src/Qorus/test/qlib/QUnit.qm:457 (Qore user code)
  call_function_args() called at /Users/david/src/Qorus/test/qlib/QUnit.qm:457 (Qore builtin code)
  TestCase::run() called at /Users/david/src/Qorus/test/qlib/QUnit.qm:2008 (Qore user code)
  Test::main() called at ../test/jni.qtest:122 (Qore user code)
  Main::constructor() called at :-1 (Qore user code)
```

File: include/qore/ExceptionSink.h [hunk 1]
```diff
@@ -309,30 +309,176 @@ static inline void makeAccessDeletedObjectException(ExceptionSink *xsink, const
...
+        const char* source = nullptr, int offset = 0, const char* lang = nullptr);
 
+    //! returns the source location
     DLLEXPORT const QoreProgramLocation& get() const {
         return *loc;
     }
 
+    //! returns the file name
+    DLLEXPORT const char* getFile() const {
+        return file_str.c_str();
```
Comment (sejvlond) [inline include/qore/ExceptionSink.h]: why returning `const char *` instead of `string`?

File: include/qore/ExceptionSink.h [hunk 2]
```diff
@@ -309,30 +309,176 @@ static inline void makeAccessDeletedObjectException(ExceptionSink *xsink, const
...
+    DLLLOCAL QoreStackLocation(QoreStackLocation&&) = default;
+
+    //! virtual destructor
+    DLLLOCAL virtual ~QoreStackLocation() = default;
+
+    //! no assignment operator
+    DLLLOCAL QoreStackLocation& operator=(const QoreStackLocation&) = delete;
+
+    //! no move assignment operator
+    DLLLOCAL QoreStackLocation& operator=(QoreStackLocation&&) = delete;
```
Comment (sejvlond) [inline include/qore/ExceptionSink.h]: what's the reason for deleting copy and move =operators? it looks that the member pointer is not owned and copies are just shallow copies..

File: include/qore/ExceptionSink.h [hunk 3]
```diff
@@ -309,30 +309,176 @@ static inline void makeAccessDeletedObjectException(ExceptionSink *xsink, const
...
+
+    //! called when pushed on the stack to set the next location
+    DLLLOCAL void setNext(const QoreStackLocation* next) {
```
Comment (sejvlond) [inline include/qore/ExceptionSink.h:424]: this methods assumes that the QoreStackLocation `next` will be valid after setNext as long as this class instance exists. But since QoreStackLocation is a simple object holding just a pointer we can copy it and we don't have to be careful that the QoreStackLocation `next` will be there as long as needed.

**Comment by davidnich** [comment]
now a native Java example:
```
QUnit Test "jni test" v1.0
System exception: NativeMethodAccessorImpl.java:-2 (Java): JNI-ERROR: java.lang.reflect.InvocationTargetException (arg: {<InvocationTargetException object>})
call stack:
  java.lang.reflect.Method.invoke() called at Method.java:498 (Java user code)
  sun.reflect.DelegatingMethodAccessorImpl.invoke() called at DelegatingMethodAccessorImpl.java:43 (Java user code)
  sun.reflect.NativeMethodAccessorImpl.invoke() called at NativeMethodAccessorImpl.java:62 (Java user code)
  Method::invoke() called at ../test/jni.qtest:722 (Qore builtin code)
  Main::testCallback() called at /Users/david/src/Qorus/test/qlib/QUnit.qm:457 (Qore user code)
  call_function_args() called at /Users/david/src/Qorus/test/qlib/QUnit.qm:457 (Qore builtin code)
  TestCase::run() called at /Users/david/src/Qorus/test/qlib/QUnit.qm:2008 (Qore user code)
  Test::main() called at ../test/jni.qtest:122 (Qore user code)
  Main::constructor() called at :-1 (Qore user code)
Ran 23 test cases, 23 succeeded (212 assertions)
```

and Qore from Java from Qore:
```
System exception: ../test/jni.qtest:127 (Qore): NO-FUNCTION: function name 'does_not_exist' does not exist
call stack:
  org.qore.jni.QoreJavaApi.callFunction0() called at QoreJavaApi.java:-2 (Java builtin code)
  org.qore.jni.QoreJavaApi.callFunction() called at QoreJavaApi.java:15 (Java user code)
  org.qore.jni.test.QoreJavaApiTest.callFunctionTest() called at QoreJavaApiTest.java:94 (Java user code)
  QoreJavaApiTest::callFunctionTest() called at ../test/jni.qtest:127 (Qore builtin code)
  Main::exceptionStackTest() called at /Users/david/src/Qorus/test/qlib/QUnit.qm:457 (Qore user code)
  call_function_args() called at /Users/david/src/Qorus/test/qlib/QUnit.qm:457 (Qore builtin code)
  TestCase::run() called at /Users/david/src/Qorus/test/qlib/QUnit.qm:2008 (Qore user code)
  Test::main() called at ../test/jni.qtest:122 (Qore user code)
  Main::constructor() called at :-1 (Qore user code)
```

**Comment by davidnich** [comment]
@sejvlond now this is ready for review - call stack info is only created when necessary - when there is a request for the call stack or an exception is raised.  Modules providing language integration (ex `jni`) add their stack info dynamically on demand.  There are no more heap allocations made for each call in Qore to maintain a separate heap-allocated call stack - the call stack is stored on each thread's stack.

Also a `lang` key has been added to the exception info + call stack entries.  

The default exception and warning handlers share the same code for formatting the output - it's in one place now, so if we want to rework it to make it more easily machine parsable, we can do so.

Looking forward to your c++ comments :)

File: include/qore/ExceptionSink.h [hunk 4]
```diff
@@ -309,30 +331,176 @@ static inline void makeAccessDeletedObjectException(ExceptionSink *xsink, const
...
+    std::string lang_str;
 
     // actual exception location
     QoreProgramLocation* loc;
 };
 
+//! Stack location element abstract class
+/** @since %Qore 0.9
+*/
+class QoreStackLocation {
```
Comment (sejvlond) [inline include/qore/ExceptionSink.h:396]: this class looks like an implementation of a linked list, maybe we can use std::list for it and have QoreStackLocation as a template class for it holding only info about the current Program, statement etc, not holding the list directly.

File: include/qore/intern/qore_thread_intern.h [hunk 5]
```diff
@@ -418,40 +412,98 @@ class QoreParseClassHelper {
...
+protected:
+    const AbstractStatement* stmt;
+    QoreProgram* pgm;
 };
 
-class QoreProgramOptionalLocationHelper {
+class QoreInternalCallStackLocationHelper : public QoreInternalCallStackLocationHelperBase {
+public:
+    DLLLOCAL QoreInternalCallStackLocationHelper(const QoreProgramLocation& loc, const char* call,
+        qore_call_t call_type) : loc(loc), call(call), call_type(call_type) {
```
Comment (sejvlond) [inline include/qore/intern/qore_thread_intern.h:451]: `call` `const char *` is stored into a member but who is the owner and how it is forced that the pointer is going to be valid as long as this instance lives?

File: include/qore/intern/qore_thread_intern.h [hunk 6]
```diff
@@ -418,40 +412,98 @@ class QoreParseClassHelper {
    DLLLOCAL ~QoreParseClassHelper();
 };
 
-class QoreProgramLocationHelper {
+class QoreProgramStackLocationHelper {
+public:
+    DLLLOCAL QoreProgramStackLocationHelper(QoreStackLocation* stack_loc, const AbstractStatement*& current_stmt,
+        QoreProgram*& current_pgm) :
```
Comment (sejvlond) [inline include/qore/intern/qore_thread_intern.h:418]: calling the ctor of `QoreProgramStackLocationHelper` changes the `current_stmt` and `current_pgm`? I don't get it why?

File: include/qore/ExceptionSink.h (hunk 3 above)
Comment (davidnich) [reply include/qore/ExceptionSink.h:424]: I don't understand the problem or how to solve it - can you please give me more info?
`next` is a pointer that must be valid while the current object is valid, because it's the current last stack location.  It will only disappear after the current item has been deleted.  If there is a better way to do this, please let me know.

File: include/qore/ExceptionSink.h (hunk 4 above)
Comment (davidnich) [reply include/qore/ExceptionSink.h:396]: The reason it's implemented like this with a virtual function to return the next stack location is so that language modules (ex `jni`) can put a single placeholder on the stack, and then if and when an exception is raised, the object can populate all the stack entries in the language stack, the number of which entries will only be known at runtime in case of an exception.

The function returning the next ptr looks like this:
```
const QoreStackLocation* QoreJniStackLocationHelper::getNext() const {
    checkInit();
    assert(current < size);
    return (++current < size) ? this : stack_next;
}
```

So basically the "substack" of Java stack entries is created on demand if the stack is actually traversed due to an exception.

File: include/qore/intern/qore_thread_intern.h (hunk 6 above)
Comment (davidnich) [reply include/qore/intern/qore_thread_intern.h:418]: it retrieves the current program and statement so they can be returned by the the `QoreStackLocation` using that object in its stack position - this functionality is used by the debugger.

File: include/qore/ExceptionSink.h (hunk 3 above)
Comment (sejvlond) [reply include/qore/ExceptionSink.h:424]: who is the owner of the pointer and when it gets deleted? because the API of this method takes a const* - readonly pointer somewhere - there are no guaranties that the pointer will be valid and still the method saves it into a member and accesses it later. So either the `QoreStackLocation` class is going to be owner of it, then I would do `setNext(std::unique_ptr<const QoreStackLocation> next)` or is not, and then it must be internally copied.

otherwise someone can do
```
QoreStackLocation loc1;
{
  QoreStackLocation loc2;
  loc1.setNext(&loc2);
}
loc1.next().smth - FAIL
```

Comment (sejvlond) [reply include/qore/ExceptionSink.h:424]: the same for the member right, it's just a pointer without knowing who owns it, I would go with unique or shared_ptr probably...?
Or as I said in a different comment, maybe the StackLocation should hold only info about the location itself and the linked list should be done by std::list - single responsibility for the object.

File: include/qore/ExceptionSink.h (hunk 2 above)
Comment (davidnich) [reply include/qore/ExceptionSink.h]: done

File: include/qore/intern/qore_thread_intern.h (hunk 5 above)
Comment (davidnich) [reply include/qore/intern/qore_thread_intern.h:451]: done

File: include/qore/ExceptionSink.h (hunk 1 above)
Comment (davidnich) [reply include/qore/ExceptionSink.h]: done
